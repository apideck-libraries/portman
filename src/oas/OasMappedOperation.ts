import { OpenAPIV3 } from 'openapi-types'

export interface IOasMappedOperation {
  id?: string
  path: string
  method: string
  pathRef: string
  schema: OpenAPIV3.OperationObject
  requestHeaders: OpenAPIV3.ParameterObject[]
  pathParams: OpenAPIV3.ParameterObject[]
  queryParams: OpenAPIV3.ParameterObject[]
  cookieParams: OpenAPIV3.ParameterObject[]

  requestBody(): OpenAPIV3.RequestBodyObject | undefined
}

export class OasMappedOperation implements IOasMappedOperation {
  public id?: string
  public path: string
  public method: string
  public pathRef: string
  public requestHeaders: OpenAPIV3.ParameterObject[]
  public pathParams: OpenAPIV3.ParameterObject[]
  public queryParams: OpenAPIV3.ParameterObject[]
  public cookieParams: OpenAPIV3.ParameterObject[]
  public schema: OpenAPIV3.OperationObject

  constructor(path: string, method: string, operation: OpenAPIV3.OperationObject) {
    this.schema = { ...operation }
    this.id = this.schema?.operationId
    this.method = method.toUpperCase()
    this.path = path
    this.pathRef = `${this.method}::${path}`
    this.requestHeaders = this.mapParameters('header')
    this.pathParams = this.mapParameters('path')
    this.queryParams = this.mapParameters('query')
    this.cookieParams = this.mapParameters('cookie')
  }

  private mapParameters(paramIn: string): OpenAPIV3.ParameterObject[] {
    if (!this.schema?.parameters) return []

    return (
      (this.schema.parameters as OpenAPIV3.ParameterObject[]).filter(
        parameter => (parameter.in as string) === paramIn
      ) || []
    )
  }

  requestBody(): OpenAPIV3.RequestBodyObject | undefined {
    return this.schema?.requestBody as OpenAPIV3.RequestBodyObject
  }

  requestBodySchema(mediaType: string): OpenAPIV3.SchemaObject | undefined {
    const requestBody = this.requestBody()
    if (!requestBody || !requestBody?.content?.[mediaType]?.schema) return

    return requestBody.content[mediaType].schema as OpenAPIV3.SchemaObject
  }
}
