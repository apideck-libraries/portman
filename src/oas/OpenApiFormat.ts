import { OpenAPIV3 } from 'openapi-types'
import openapiFormat, { OpenAPIFilterOptions, OpenAPIFilterSet } from 'openapi-format'

export interface IOpenApiFormatterConfig {
  inputFile: string
  filterFile: string
  outputFile?: string
}

export class OpenApiFormatter {
  public oas: OpenAPIV3.Document

  async filter(options: IOpenApiFormatterConfig): Promise<OpenAPIV3.Document> {
    const inputFilePath = options.inputFile
    const filterFilePath = options.filterFile
    const filterOptions = {} as OpenAPIFilterOptions

    // Load OpenAPI file
    this.oas = (await openapiFormat.parseFile(inputFilePath)) as unknown as OpenAPIV3.Document

    // Load filter file
    const filterSet = (await openapiFormat.parseFile(filterFilePath)) as OpenAPIFilterSet
    filterOptions.filterSet = filterSet

    // Filter OpenAPI file
    const resFilter = await openapiFormat.openapiFilter(this.oas, filterOptions)
    this.oas = resFilter.data as OpenAPIV3.Document

    // Write filtered OpenAPI file as YAML
    if (options.outputFile) {
      await openapiFormat.writeFile(options.outputFile, this.oas, { format: 'yaml' })
    }

    return this.oas
  }

  async parseFile(filePath: string): Promise<Record<string, unknown>> {
    return await openapiFormat.parseFile(filePath)
  }

  changeCase(valueAsString: string, caseType: string): string {
    return openapiFormat.changeCase(valueAsString, caseType)
  }
}
