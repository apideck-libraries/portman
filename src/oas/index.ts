export * from './OasMappedOperation'
export * from './OpenApiParser'
