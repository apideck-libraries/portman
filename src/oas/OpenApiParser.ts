import SwaggerParser from '@apidevtools/swagger-parser'
import { OpenAPIV3 } from 'openapi-types'
import path from 'path'
import { matchPath, METHODS } from 'utils'
import { OasMappedOperation } from './OasMappedOperation'

export interface OpenApiParserConfig {
  inputFile: string
}

export interface IOpenApiParser {
  oas: OpenAPIV3.Document
  mappedOperations: OasMappedOperation[]
  operationIdMap: Record<string, OasMappedOperation>
  getOperationById(operationId: string): OasMappedOperation | null
  getOperationByPath(path: string): OasMappedOperation | null
}

export class OpenApiParser {
  public oas: OpenAPIV3.Document
  public mappedOperations: OasMappedOperation[]
  public operationIdMap: Record<string, OasMappedOperation>

  async convert(options: OpenApiParserConfig): Promise<OpenAPIV3.Document> {
    const inputFile = path.resolve(options.inputFile)
    // Dereference the spec so all entities have encapsulated schemas
    const api = await SwaggerParser.dereference(inputFile)
    const oasDoc = (await SwaggerParser.bundle(api, {
      parse: {
        json: false, // Disable the JSON parser
        yaml: {
          allowEmpty: false // Don't allow empty YAML files
        }
      },
      dereference: {
        circular: false // Don't allow circular $refs
      },
      validate: {
        spec: false // Don't validate against the Swagger spec
      }
    })) as OpenAPIV3.Document

    this.oas = oasDoc
    this.pathsToOperations()
    this.mapOperationIds()
    return oasDoc
  }

  pathsToOperations = (): OasMappedOperation[] => {
    const paths = this.oas.paths
    const mappedOperations = Object.entries(paths)
      .filter(([_path, operations]) => !!operations)
      .map(([path, operations]) => {
        return (
          operations &&
          Object.entries(operations).map(([method, operation]): OasMappedOperation => {
            return new OasMappedOperation(path, method, operation as OpenAPIV3.OperationObject)
          })
        )
      })
      .reduce<OasMappedOperation[]>((acc, resource) => {
        resource && resource.map(item => acc.push(item))
        return acc
      }, [])

    this.mappedOperations = mappedOperations
    return this.mappedOperations
  }

  mapOperationIds = (): void => {
    const operationIdMap = this.mappedOperations.reduce((acc, operation) => {
      acc[operation.pathRef] = operation
      return acc
    }, {})

    this.operationIdMap = operationIdMap
  }

  public getOperationById(operationId: string): OasMappedOperation | null {
    return this.mappedOperations.find(({ id }) => id === operationId) || null
  }

  public getOperationByPath(path: string): OasMappedOperation | null {
    return this.mappedOperations.find(({ pathRef }) => pathRef === path) || null
  }

  public getOperationsByIds(operationIds: string[]): OasMappedOperation[] {
    return this.mappedOperations.filter(({ id }) => id && operationIds.includes(id))
  }

  public getOperationsByPath(path: string): OasMappedOperation[] {
    const targetSplit = path.split('::')
    const targetMethod = targetSplit[0].includes('*') ? METHODS : targetSplit[0]
    const targetPath = targetSplit[1]

    return this.mappedOperations.filter(mappedOperation => {
      return (
        targetMethod.includes(mappedOperation.method) && matchPath(targetPath, mappedOperation.path)
      )
    })
  }
}
