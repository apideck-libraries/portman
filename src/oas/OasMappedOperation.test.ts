import { OasMappedOperation, OpenApiParser } from 'oas'
import { OpenAPIV3 } from 'openapi-types'

describe('OasMappedOperation', () => {
  let specOperation: OpenAPIV3.OperationObject
  let mappedOperation: OasMappedOperation

  const oasYml = '__tests__/fixtures/crm.yml'
  const parser = new OpenApiParser()

  const path = '/crm/companies'
  const method = 'post'

  beforeEach(async () => {
    await parser.convert({ inputFile: oasYml })
    const paths = parser.oas?.paths
    expect(paths).toBeDefined()

    specOperation = Object.values(paths)?.[0]?.['post'] || {}
    expect(specOperation).toBeDefined()
    mappedOperation = new OasMappedOperation(path, method, specOperation)
  })

  it(`should set itself up using operation and params`, () => {
    expect(mappedOperation).toMatchSnapshot()
  })

  it('should handle operationId not present', () => {
    const { operationId, ...operation } = specOperation
    mappedOperation = new OasMappedOperation(path, method, operation)
    expect(mappedOperation.id).not.toBeDefined()
  })

  describe('requestBodySchema', () => {
    it('should have access to the requestBody for mediaType', () => {
      const mediaType = 'application/json'
      const requestBody = mappedOperation.requestBodySchema(mediaType)
      expect(requestBody).toMatchSnapshot()
    })

    it('should return null for unknown mediaType', () => {
      const mediaType = 'application/xml'
      const requestBodySchema = mappedOperation.requestBodySchema(mediaType)
      expect(requestBodySchema).not.toBeDefined()
    })

    it('should return null when no requestBody schema exists', () => {
      const mediaType = 'application/xml'
      const { requestBody, ...operation } = specOperation
      mappedOperation = new OasMappedOperation(path, method, operation)

      const requestBodySchema = mappedOperation.requestBodySchema(mediaType)
      expect(requestBodySchema).not.toBeDefined()
    })
  })
})
