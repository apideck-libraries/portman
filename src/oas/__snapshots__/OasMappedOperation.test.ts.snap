// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OasMappedOperation requestBodySchema should have access to the requestBody for mediaType 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "addresses": Object {
      "items": Object {
        "properties": Object {
          "city": Object {
            "description": "Name of city.",
            "example": "San Francisco",
            "nullable": true,
            "type": "string",
          },
          "country": Object {
            "description": "country code according to ISO 3166-1 alpha-2.",
            "example": "US",
            "nullable": true,
            "type": "string",
          },
          "id": Object {
            "example": "123",
            "nullable": true,
            "type": "string",
          },
          "latitude": Object {
            "example": "40.759211",
            "nullable": true,
            "type": "string",
          },
          "line1": Object {
            "description": "Line 1 of the address e.g. number, street, suite, apt #, etc.",
            "example": "Main street",
            "nullable": true,
            "type": "string",
          },
          "line2": Object {
            "description": "Line 2 of the address",
            "example": "apt #",
            "nullable": true,
            "type": "string",
          },
          "longitude": Object {
            "example": "-73.984638",
            "nullable": true,
            "type": "string",
          },
          "name": Object {
            "example": "HQ US",
            "nullable": true,
            "type": "string",
          },
          "postal_code": Object {
            "description": "Zip code or equivalent.",
            "example": "94104",
            "nullable": true,
            "type": "string",
          },
          "state": Object {
            "description": "Name of state",
            "example": "CA",
            "nullable": true,
            "type": "string",
          },
          "type": Object {
            "enum": Array [
              "primary",
              "secondary",
              "home",
              "office",
              "shipping",
              "billing",
              "other",
            ],
            "example": "primary",
            "type": "string",
            "x-graphql-type-name": "AddressType",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "bank_accounts": Object {
      "items": Object {
        "properties": Object {
          "bic": Object {
            "example": "AUDSCHGGXXX",
            "nullable": true,
            "type": "string",
          },
          "iban": Object {
            "example": "CH2989144532982975332",
            "nullable": true,
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "created_at": Object {
      "example": "2020-09-30T07:43:32.000Z",
      "readOnly": true,
      "type": "string",
    },
    "created_by": Object {
      "example": "12345",
      "nullable": true,
      "readOnly": true,
      "type": "string",
    },
    "currency": Object {
      "example": "USD",
      "nullable": true,
      "type": "string",
    },
    "custom_fields": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "id": Object {
            "example": "custom_technologies",
            "type": "string",
          },
          "value": Object {
            "anyOf": Array [
              Object {
                "example": "Uses Salesforce and Marketo",
                "nullable": true,
                "type": "string",
              },
              Object {
                "example": 10,
                "nullable": true,
                "type": "number",
              },
              Object {
                "example": true,
                "nullable": true,
                "type": "boolean",
              },
              Object {
                "items": Object {
                  "type": "string",
                },
                "type": "array",
              },
            ],
          },
        },
        "required": Array [
          "id",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "description": Object {
      "example": "A crm that works for you, so you can spend time on relationships instead of data.",
      "nullable": true,
      "type": "string",
    },
    "emails": Object {
      "items": Object {
        "properties": Object {
          "email": Object {
            "example": "elon@musk.com",
            "format": "email",
            "minLength": 1,
            "type": "string",
          },
          "id": Object {
            "example": "123",
            "type": "string",
          },
          "type": Object {
            "enum": Array [
              "primary",
              "secondary",
              "work",
              "personal",
              "billing",
              "other",
            ],
            "example": "primary",
            "type": "string",
            "x-graphql-type-name": "EmailType",
          },
        },
        "required": Array [
          "email",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "fax": Object {
      "example": "+12129876543",
      "nullable": true,
      "type": "string",
    },
    "id": Object {
      "example": "12345",
      "readOnly": true,
      "type": "string",
    },
    "image_url": Object {
      "example": "https://www.spacex.com/static/images/share.jpg",
      "nullable": true,
      "type": "string",
    },
    "interaction_count": Object {
      "example": 1,
      "nullable": true,
      "readOnly": true,
      "type": "integer",
    },
    "name": Object {
      "example": "Copper",
      "minLength": 1,
      "type": "string",
    },
    "owner_id": Object {
      "example": "12345",
      "type": "string",
    },
    "phone_numbers": Object {
      "items": Object {
        "properties": Object {
          "id": Object {
            "example": "12345",
            "nullable": true,
            "type": "string",
          },
          "number": Object {
            "example": "111-111-1111",
            "minLength": 1,
            "type": "string",
          },
          "type": Object {
            "enum": Array [
              "primary",
              "secondary",
              "home",
              "office",
              "mobile",
              "assistant",
              "fax",
              "other",
            ],
            "example": "primary",
            "type": "string",
            "x-graphql-type-name": "PhoneType",
          },
        },
        "required": Array [
          "number",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "social_links": Object {
      "items": Object {
        "properties": Object {
          "id": Object {
            "example": "12345",
            "nullable": true,
            "type": "string",
          },
          "type": Object {
            "example": "twitter",
            "nullable": true,
            "type": "string",
          },
          "url": Object {
            "example": "https://www.twitter.com/apideck-io",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "url",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "status": Object {
      "example": "Open",
      "minLength": 1,
      "nullable": true,
      "type": "string",
    },
    "tags": Object {
      "example": Array [
        "New",
      ],
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
    "updated_at": Object {
      "example": "2020-09-30T07:43:32.000Z",
      "readOnly": true,
      "type": "string",
    },
    "updated_by": Object {
      "example": "12345",
      "nullable": true,
      "readOnly": true,
      "type": "string",
    },
    "vat_number": Object {
      "description": "VAT number",
      "example": "BE0689615164",
      "nullable": true,
      "type": "string",
    },
    "websites": Object {
      "items": Object {
        "properties": Object {
          "id": Object {
            "example": "12345",
            "nullable": true,
            "type": "string",
          },
          "type": Object {
            "enum": Array [
              "primary",
              "secondary",
              "work",
              "personal",
              "other",
            ],
            "example": "primary",
            "type": "string",
            "x-graphql-type-name": "WebsiteType",
          },
          "url": Object {
            "example": "http://example.com",
            "minLength": 1,
            "type": "string",
          },
        },
        "required": Array [
          "url",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": Array [
    "name",
  ],
  "type": "object",
}
`;

exports[`OasMappedOperation should set itself up using operation and params 1`] = `
OasMappedOperation {
  "cookieParams": Array [],
  "id": "companiesAdd",
  "method": "POST",
  "path": "/crm/companies",
  "pathParams": Array [],
  "pathRef": "POST::/crm/companies",
  "queryParams": Array [
    Object {
      "description": "Include raw response. Mostly used for debugging purposes",
      "in": "query",
      "name": "raw",
      "schema": Object {
        "default": true,
        "type": "boolean",
      },
    },
  ],
  "requestHeaders": Array [
    Object {
      "description": "ID of the consumer which you want to get or push data from",
      "in": "header",
      "name": "x-apideck-consumer-id",
      "required": true,
      "schema": Object {
        "type": "string",
      },
    },
    Object {
      "description": "The ID of your Unify application",
      "in": "header",
      "name": "x-apideck-app-id",
      "required": true,
      "schema": Object {
        "type": "string",
      },
    },
    Object {
      "description": "Provide the service id you want to call (e.g., pipedrive). [See the full list in the connector section.](#section/Connectors) Only needed when a consumer has activated multiple integrations for a Unified API.",
      "in": "header",
      "name": "x-apideck-service-id",
      "schema": Object {
        "type": "string",
      },
    },
  ],
  "responseCodes": Array [
    "201",
    "400",
    "401",
    "402",
    "404",
    "422",
    "default",
  ],
  "schema": Object {
    "description": "Create company",
    "operationId": "companiesAdd",
    "parameters": Array [
      Object {
        "description": "Include raw response. Mostly used for debugging purposes",
        "in": "query",
        "name": "raw",
        "schema": Object {
          "default": true,
          "type": "boolean",
        },
      },
      Object {
        "description": "ID of the consumer which you want to get or push data from",
        "in": "header",
        "name": "x-apideck-consumer-id",
        "required": true,
        "schema": Object {
          "type": "string",
        },
      },
      Object {
        "description": "The ID of your Unify application",
        "in": "header",
        "name": "x-apideck-app-id",
        "required": true,
        "schema": Object {
          "type": "string",
        },
      },
      Object {
        "description": "Provide the service id you want to call (e.g., pipedrive). [See the full list in the connector section.](#section/Connectors) Only needed when a consumer has activated multiple integrations for a Unified API.",
        "in": "header",
        "name": "x-apideck-service-id",
        "schema": Object {
          "type": "string",
        },
      },
    ],
    "requestBody": Object {
      "content": Object {
        "application/json": Object {
          "schema": Object {
            "additionalProperties": false,
            "properties": Object {
              "addresses": Object {
                "items": Object {
                  "properties": Object {
                    "city": Object {
                      "description": "Name of city.",
                      "example": "San Francisco",
                      "nullable": true,
                      "type": "string",
                    },
                    "country": Object {
                      "description": "country code according to ISO 3166-1 alpha-2.",
                      "example": "US",
                      "nullable": true,
                      "type": "string",
                    },
                    "id": Object {
                      "example": "123",
                      "nullable": true,
                      "type": "string",
                    },
                    "latitude": Object {
                      "example": "40.759211",
                      "nullable": true,
                      "type": "string",
                    },
                    "line1": Object {
                      "description": "Line 1 of the address e.g. number, street, suite, apt #, etc.",
                      "example": "Main street",
                      "nullable": true,
                      "type": "string",
                    },
                    "line2": Object {
                      "description": "Line 2 of the address",
                      "example": "apt #",
                      "nullable": true,
                      "type": "string",
                    },
                    "longitude": Object {
                      "example": "-73.984638",
                      "nullable": true,
                      "type": "string",
                    },
                    "name": Object {
                      "example": "HQ US",
                      "nullable": true,
                      "type": "string",
                    },
                    "postal_code": Object {
                      "description": "Zip code or equivalent.",
                      "example": "94104",
                      "nullable": true,
                      "type": "string",
                    },
                    "state": Object {
                      "description": "Name of state",
                      "example": "CA",
                      "nullable": true,
                      "type": "string",
                    },
                    "type": Object {
                      "enum": Array [
                        "primary",
                        "secondary",
                        "home",
                        "office",
                        "shipping",
                        "billing",
                        "other",
                      ],
                      "example": "primary",
                      "type": "string",
                      "x-graphql-type-name": "AddressType",
                    },
                  },
                  "type": "object",
                },
                "type": "array",
              },
              "bank_accounts": Object {
                "items": Object {
                  "properties": Object {
                    "bic": Object {
                      "example": "AUDSCHGGXXX",
                      "nullable": true,
                      "type": "string",
                    },
                    "iban": Object {
                      "example": "CH2989144532982975332",
                      "nullable": true,
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
                "type": "array",
              },
              "created_at": Object {
                "example": "2020-09-30T07:43:32.000Z",
                "readOnly": true,
                "type": "string",
              },
              "created_by": Object {
                "example": "12345",
                "nullable": true,
                "readOnly": true,
                "type": "string",
              },
              "currency": Object {
                "example": "USD",
                "nullable": true,
                "type": "string",
              },
              "custom_fields": Object {
                "items": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "id": Object {
                      "example": "custom_technologies",
                      "type": "string",
                    },
                    "value": Object {
                      "anyOf": Array [
                        Object {
                          "example": "Uses Salesforce and Marketo",
                          "nullable": true,
                          "type": "string",
                        },
                        Object {
                          "example": 10,
                          "nullable": true,
                          "type": "number",
                        },
                        Object {
                          "example": true,
                          "nullable": true,
                          "type": "boolean",
                        },
                        Object {
                          "items": Object {
                            "type": "string",
                          },
                          "type": "array",
                        },
                      ],
                    },
                  },
                  "required": Array [
                    "id",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "description": Object {
                "example": "A crm that works for you, so you can spend time on relationships instead of data.",
                "nullable": true,
                "type": "string",
              },
              "emails": Object {
                "items": Object {
                  "properties": Object {
                    "email": Object {
                      "example": "elon@musk.com",
                      "format": "email",
                      "minLength": 1,
                      "type": "string",
                    },
                    "id": Object {
                      "example": "123",
                      "type": "string",
                    },
                    "type": Object {
                      "enum": Array [
                        "primary",
                        "secondary",
                        "work",
                        "personal",
                        "billing",
                        "other",
                      ],
                      "example": "primary",
                      "type": "string",
                      "x-graphql-type-name": "EmailType",
                    },
                  },
                  "required": Array [
                    "email",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "fax": Object {
                "example": "+12129876543",
                "nullable": true,
                "type": "string",
              },
              "id": Object {
                "example": "12345",
                "readOnly": true,
                "type": "string",
              },
              "image_url": Object {
                "example": "https://www.spacex.com/static/images/share.jpg",
                "nullable": true,
                "type": "string",
              },
              "interaction_count": Object {
                "example": 1,
                "nullable": true,
                "readOnly": true,
                "type": "integer",
              },
              "name": Object {
                "example": "Copper",
                "minLength": 1,
                "type": "string",
              },
              "owner_id": Object {
                "example": "12345",
                "type": "string",
              },
              "phone_numbers": Object {
                "items": Object {
                  "properties": Object {
                    "id": Object {
                      "example": "12345",
                      "nullable": true,
                      "type": "string",
                    },
                    "number": Object {
                      "example": "111-111-1111",
                      "minLength": 1,
                      "type": "string",
                    },
                    "type": Object {
                      "enum": Array [
                        "primary",
                        "secondary",
                        "home",
                        "office",
                        "mobile",
                        "assistant",
                        "fax",
                        "other",
                      ],
                      "example": "primary",
                      "type": "string",
                      "x-graphql-type-name": "PhoneType",
                    },
                  },
                  "required": Array [
                    "number",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "social_links": Object {
                "items": Object {
                  "properties": Object {
                    "id": Object {
                      "example": "12345",
                      "nullable": true,
                      "type": "string",
                    },
                    "type": Object {
                      "example": "twitter",
                      "nullable": true,
                      "type": "string",
                    },
                    "url": Object {
                      "example": "https://www.twitter.com/apideck-io",
                      "minLength": 1,
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "url",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "status": Object {
                "example": "Open",
                "minLength": 1,
                "nullable": true,
                "type": "string",
              },
              "tags": Object {
                "example": Array [
                  "New",
                ],
                "items": Object {
                  "type": "string",
                },
                "type": "array",
              },
              "updated_at": Object {
                "example": "2020-09-30T07:43:32.000Z",
                "readOnly": true,
                "type": "string",
              },
              "updated_by": Object {
                "example": "12345",
                "nullable": true,
                "readOnly": true,
                "type": "string",
              },
              "vat_number": Object {
                "description": "VAT number",
                "example": "BE0689615164",
                "nullable": true,
                "type": "string",
              },
              "websites": Object {
                "items": Object {
                  "properties": Object {
                    "id": Object {
                      "example": "12345",
                      "nullable": true,
                      "type": "string",
                    },
                    "type": Object {
                      "enum": Array [
                        "primary",
                        "secondary",
                        "work",
                        "personal",
                        "other",
                      ],
                      "example": "primary",
                      "type": "string",
                      "x-graphql-type-name": "WebsiteType",
                    },
                    "url": Object {
                      "example": "http://example.com",
                      "minLength": 1,
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "url",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
            },
            "required": Array [
              "name",
            ],
            "type": "object",
          },
        },
      },
      "required": true,
    },
    "responses": Object {
      "201": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "properties": Object {
                    "id": Object {
                      "example": "12345",
                      "readOnly": true,
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "id",
                  ],
                  "title": "UnifiedId",
                },
                "operation": Object {
                  "description": "Operation performed",
                  "example": "one",
                  "type": "string",
                },
                "resource": Object {
                  "description": "Unified API resource name",
                  "example": "companies",
                  "type": "string",
                },
                "service": Object {
                  "description": "Apideck ID of service provider",
                  "example": "zoho-crm",
                  "type": "string",
                },
                "status": Object {
                  "description": "HTTP Response Status",
                  "example": "OK",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP Response Status Code",
                  "example": 200,
                  "type": "integer",
                },
              },
              "required": Array [
                "status_code",
                "status",
                "service",
                "resource",
                "operation",
                "data",
              ],
              "type": "object",
            },
          },
        },
        "description": "Company created",
      },
      "400": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "detail": Object {
                  "anyOf": Array [
                    Object {
                      "example": "Missing property foobar",
                      "type": "string",
                    },
                    Object {
                      "example": Object {
                        "missing": Array [
                          Array [
                            Object {
                              "foobar": "required",
                            },
                          ],
                        ],
                      },
                      "type": "object",
                    },
                  ],
                  "description": "Contains parameter or domain specific information related to the error and why it occured.",
                },
                "error": Object {
                  "description": "Contains an explanation of the status_code as defined in HTTP/1.1 standard (RFC 7231)",
                  "example": "Bad Request",
                  "type": "string",
                },
                "message": Object {
                  "description": "A human-readable message providing more details about the error.",
                  "example": "Invalid Params",
                  "type": "string",
                },
                "ref": Object {
                  "description": "Link to documentation of error type",
                  "example": "https://developers.apideck.com/errors#requestbodyvalidationerror",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP status code",
                  "example": 400,
                  "type": "number",
                },
                "typeName": Object {
                  "description": "The type of error returned",
                  "example": "RequestHeadersValidationError",
                  "type": "string",
                },
              },
            },
          },
        },
        "description": "Bad Request",
      },
      "401": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "detail": Object {
                  "description": "Contains parameter or domain specific information related to the error and why it occured.",
                  "example": "Failed to generate valid JWT Session. Verify applicationId is correct",
                  "type": "string",
                },
                "error": Object {
                  "description": "Contains an explanation of the status_code as defined in HTTP/1.1 standard (RFC 7231)",
                  "example": "Unauthorized",
                  "type": "string",
                },
                "message": Object {
                  "description": "A human-readable message providing more details about the error.",
                  "example": "Unauthorized Request",
                  "type": "string",
                },
                "ref": Object {
                  "description": "Link to documentation of error type",
                  "example": "https://developers.apideck.com/errors#unauthorizederror",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP status code",
                  "example": 401,
                  "type": "number",
                },
                "typeName": Object {
                  "description": "The type of error returned",
                  "example": "UnauthorizedError",
                  "type": "string",
                },
              },
            },
          },
        },
        "description": "Unauthorized",
      },
      "402": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "detail": Object {
                  "description": "Contains parameter or domain specific information related to the error and why it occured.",
                  "example": "You have reached your limit of 2000",
                  "type": "string",
                },
                "error": Object {
                  "description": "Contains an explanation of the status_code as defined in HTTP/1.1 standard (RFC 7231)",
                  "example": "Payment Required",
                  "type": "string",
                },
                "message": Object {
                  "description": "A human-readable message providing more details about the error.",
                  "example": "Request Limit Reached",
                  "type": "string",
                },
                "ref": Object {
                  "description": "Link to documentation of error type",
                  "example": "https://developers.apideck.com/errors#requestlimiterror",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP status code",
                  "example": 402,
                  "type": "number",
                },
                "typeName": Object {
                  "description": "The type of error returned",
                  "example": "RequestLimitError",
                  "type": "string",
                },
              },
            },
          },
        },
        "description": "Payment Required",
      },
      "404": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "detail": Object {
                  "anyOf": Array [
                    Object {
                      "example": "Could not find widget with id: '123'",
                      "type": "string",
                    },
                    Object {
                      "example": Object {
                        "not_found": Object {
                          "entity": "widget",
                          "id": "123",
                        },
                      },
                      "type": "object",
                    },
                  ],
                  "description": "Contains parameter or domain specific information related to the error and why it occured.",
                },
                "error": Object {
                  "description": "Contains an explanation of the status_code as defined in HTTP/1.1 standard (RFC 7231)",
                  "example": "Not Found",
                  "type": "string",
                },
                "message": Object {
                  "description": "A human-readable message providing more details about the error.",
                  "example": "Unknown Widget",
                  "type": "string",
                },
                "ref": Object {
                  "description": "Link to documentation of error type",
                  "example": "https://developers.apideck.com/errors#entitynotfounderror",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP status code",
                  "example": 404,
                  "type": "number",
                },
                "typeName": Object {
                  "description": "The type of error returned",
                  "example": "EntityNotFoundError",
                  "type": "string",
                },
              },
            },
          },
        },
        "description": "The specified resource was not found",
      },
      "422": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "detail": Object {
                  "description": "Contains parameter or domain specific information related to the error and why it occured.",
                  "example": "State did not include unified_api",
                  "type": "string",
                },
                "error": Object {
                  "description": "Contains an explanation of the status_code as defined in HTTP/1.1 standard (RFC 7231)",
                  "example": "Unprocessable Entity",
                  "type": "string",
                },
                "message": Object {
                  "description": "A human-readable message providing more details about the error.",
                  "example": "Invalid State",
                  "type": "string",
                },
                "ref": Object {
                  "description": "Link to documentation of error type",
                  "example": "https://developers.apideck.com/errors#invalidstateerror",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP status code",
                  "example": 422,
                  "type": "number",
                },
                "typeName": Object {
                  "description": "The type of error returned",
                  "example": "InvalidStateError",
                  "type": "string",
                },
              },
            },
          },
        },
        "description": "Unprocessable",
      },
      "default": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "properties": Object {
                "detail": Object {
                  "anyOf": Array [
                    Object {
                      "example": "Missing Header: x-apideck-consumer-id",
                      "type": "string",
                    },
                    Object {
                      "example": Object {
                        "missing": Array [
                          Array [
                            Object {
                              "x-apideck-consumer-id": "required",
                            },
                          ],
                        ],
                      },
                      "type": "object",
                    },
                  ],
                  "description": "Contains parameter or domain specific information related to the error and why it occured.",
                },
                "error": Object {
                  "description": "Contains an explanation of the status_code as defined in HTTP/1.1 standard (RFC 7231)",
                  "example": "Bad Request",
                  "type": "string",
                },
                "message": Object {
                  "description": "A human-readable message providing more details about the error.",
                  "example": "Invalid Params",
                  "type": "string",
                },
                "ref": Object {
                  "description": "Link to documentation of error type",
                  "example": "https://developers.apideck.com/errors#unauthorizederror",
                  "type": "string",
                },
                "status_code": Object {
                  "description": "HTTP status code",
                  "example": 400,
                  "type": "number",
                },
                "typeName": Object {
                  "description": "The type of error returned",
                  "example": "RequestHeadersValidationError",
                  "type": "string",
                },
              },
            },
          },
        },
        "description": "Unexpected error",
      },
    },
    "security": Array [
      Object {
        "apiKey": Array [],
      },
    ],
    "summary": "Create company",
    "tags": Array [
      "Companies",
    ],
  },
}
`;
